### ISSUE TITLE
DB query binding fail on excess parameter values

### ISSUE THREAD

**Issue by webalchemist** [bcit-ci/CodeIgniter#3774]
If there are more elements in the parameter array than '?' in the query, substitutions aren't made and the query fails.

You may not regard this as an error, but it is a change in behaviour from CI 2.2.1 and should at least be noted in the upgrade docs.  In CI 2.2.1 extra parameters are simply ignored and substitutions are still made.

Example:
$sql = "SELECT * FROM my_table WHERE id=? ";
$query = $this->db->query($sql, array(1, 'anything')); // FAILS

**Comment by narfbg** [comment]
Sorry, but I don't think this is worthy of being included in the upgrade instructions (especially after the release) ... it's common sense that the count of markers and that of substitutions must be the same.

**Comment by pabloveliz** [comment]
It should behave like previous version, or get mention in the upgrade docs, or throw a better error description. 
```php
$sql = "SELECT * FROM catalogo WHERE grupo=? AND is_active";
if( $key ) {
    $sql.= " AND codigo=?";
}
$q = $this->MY->query( $sql, array($grupo, $key) );
```
This gives error on 3.0 but works fine in 2.2, at least the error should be more descriptive, and not just:
```
You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '? and is_active' at line 1
```

**Comment by hellofarhan** [comment]
Your query should be like this.

```sql
$sql = "SELECT * FROM catalogo WHERE grupo=? AND is_active=?";
if( $key ) {
    $sql.= " AND codigo=?";
}
$q = $this->MY->query( $sql, array($grupo, $active, $key) );
```

**Comment by pabloveliz** [comment]
Exactly my point, thanks.

One should change that type of code that WAS working in 2.2 but DON'T on 3.0. 

So, I have to scan all code already deployed, for this PARTICULAR case and change it (as @hellofarhan suggested) and that SHOULD GO in the upgrade notes as @webalchemist stated before, and not to expect for this particular case come and bite you in the "rear"

**Comment by webalchemist** [comment]
I guess those involved in the development of CI3 assume that (like them) all CI users have already done their upgrades and so there is no point in changing the upgrade docs after the release.  The majority of CI users (IMO) will have waited for the official release before starting their upgrades and (if they're anything like me) still have sites to convert.

Whilst I still think this should be in the upgrade docs, it doesn't look like it's going to be added. At least we've created a record of the issue and the solution here so that people can find it AFTER this issue 'bites them in the rear.'

**Comment by narfbg** [comment]
Look, the upgrade guide is quite large as it is now; and if you're experiencing this issue, it's only because you've used the feature incorrectly in the first place. Not every obscure change deserves to be documented and passing the same number of parameters as there are markers is just common sense.
